### ISSUE TITLE
Multilingual Support - adding new languages

### ISSUE THREAD

**Issue by kgschoueri** [Smithsonian/dpo-voyager#163]
Is it possible to add new languages for multilingual support? For example, Dutch and Japanese...

**Comment by gjcope** [comment]
Yes, you can add new language support to the UI by adding a language file as found here: https://github.com/Smithsonian/dpo-voyager/tree/master/assets/language and then multilingual scene content is added via the Voyager Story interface.

**Comment by kgschoueri** [comment]
Great! I will give it a try!

**Comment by gjcope** [comment]
I'm not sure if Japanese characters will be an issue, but happy to look into it if they are.

**Comment by kgschoueri** [comment]
I will try it out the Dutch first and then see if I can get the Japanese to work...

**Comment by kgschoueri** [comment]
@gjcope we weren't successful in implementing a new language in Voyager Story. We added two files string.resources.nl.json and string.resources.ja.json to the filepath app/assets/language and the server recompiled but there was no changes in our Voyager Story for new languages other than spanish and german. I had first tried adding these json files to your master source code. Would it make more sense send these files to you and have you add them to your master docs?

**Comment by gjcope** [comment]
@kgschoueri The language codes will also have to be added to the language definitions here: https://github.com/Smithsonian/dpo-voyager/blob/master/source/client/schema/common.ts 

I'm also happy to update it if you or @dirkroorda open a pull request with the files.

**Comment by dirkroorda** [comment]
@gjcope I'm glad to do that. @kgschoueri let's have a quick word about this next week, and then I test it out and make the pull request.

**Comment by dirkroorda** [comment]
I have added the changes manually to the distributed version of voyager-story. Here:

```
var EUnitType;
(function (EUnitType) {
    EUnitType[EUnitType["mm"] = 0] = "mm";
    EUnitType[EUnitType["cm"] = 1] = "cm";
    EUnitType[EUnitType["m"] = 2] = "m";
    EUnitType[EUnitType["km"] = 3] = "km";
    EUnitType[EUnitType["in"] = 4] = "in";
    EUnitType[EUnitType["ft"] = 5] = "ft";
    EUnitType[EUnitType["yd"] = 6] = "yd";
    EUnitType[EUnitType["mi"] = 7] = "mi";
})(EUnitType || (EUnitType = {}));
var ELanguageType;
(function (ELanguageType) {
    ELanguageType[ELanguageType["EN"] = 0] = "EN";
    ELanguageType[ELanguageType["ES"] = 1] = "ES";
    ELanguageType[ELanguageType["DE"] = 2] = "DE";
    ELanguageType[ELanguageType["NL"] = 3] = "NL";
    ELanguageType[ELanguageType["JA"] = 4] = "JA";
})(ELanguageType || (ELanguageType = {}));
var ELanguageStringType;
(function (ELanguageStringType) {
    ELanguageStringType["EN"] = "English";
    ELanguageStringType["ES"] = "Spanish (Espa\u00F1ol)";
    ELanguageStringType["DE"] = "German (Deutsche)";
    ELanguageStringType["NL"] = "Dutch (Nederlands)";
    ELanguageStringType["JA"] = "Japanese (日本語)";
})(ELanguageStringType || (ELanguageStringType = {}));
const DEFAULT_LANGUAGE = "EN";
```

On the interface, I can select Dutch and Japanese now, under collections.
<img width="912" alt="image" src="https://user-images.githubusercontent.com/1374704/193526699-e3a54ce0-e2ea-4777-a6b6-3a9785830813.png">

**Comment by dirkroorda** [comment]
But in doing so, the Javascript throws an error:

```
Error: failed to fetch from 'https://cdn.jsdelivr.net/gh/smithsonian/dpo-voyager@latest/assets/language/string.resources.nl.json', status: 404 
```

It seems that voyager wants its language file from the cdn. Despite the fact that the language files are locally present on the server, in `static/dist/language`.

**Comment by dirkroorda** [comment]
Even if I pass the attribute `resourceRoot="/static/dist"` I get the same behaviour.
Note that the browser parses `resourceRoot` to `resourceroot`, because HTML is supposed to be case-insensitive.
I looked up the place in the code where this attribute is used, and I changed it to 

```
resourceRoot: this.getAttribute("resourceroot")
```

but that did not help.

I think I stop here, because for me it would require a long journey in the voyager code.

**Comment by gjcope** [comment]
Hi @dirkroorda , @vicding-mi , @kgschoueri -  I have grabbed your language PR and built it locally. With the parameters set for offline use (https://smithsonian.github.io/dpo-voyager/explorer/offline/) your new language files are being supported as expected in my testing. Are you still getting console errors saying the language files could not be loaded? If not, are you able to PM me a way to access your example?

@dirkroorda Agree we should standardize the attribute names (I believe that was done for readability), but getAttribute() also lower-cases its argument before processing so this shouldn't be a functional issue.

**Comment by vicding-mi** [comment]
hi @gjcope , we are using `voyager-story` instead of `voyager-explorer`, does that make a difference? Please see the content below, how I added the resourceRoot to the right place
`# cat voyager-story-dev.html
...
    <body>
        <voyager-story resourceRoot="/assets/"></voyager-story>
    </body>
...
`

**Comment by gjcope** [comment]
Hey @vicding-mi either way, the attribute is the same. Are you using the default distribution hierarchy or have you modified it? If default, your local resourceRoot should be "./" as the distribution folder and resource folder are the same. But you should also be seeing console errors in that case. Yes, happy to take a look if needed.

**Comment by vicding-mi** [comment]
I tried again to put resourceRoot in the html file, however, they do not seem to have any effect and all the resources are still loaded through CDN. Putting `resourceroot=./` in the URL did the trick and it works well.

**Comment by gjcope** [comment]
Hi @vicding-mi - glad you got it working. Do you have an example you can share where adding the resourceRoot attribute to the component is not changing where the resources load from. It seems to be working well in tests on my end.

**Comment by kgschoueri** [comment]
@gjcope the solution is not completely resolved just yet. When I changed the url from voyager-explorer-dev to voyager-story-dev the two new languages are not showing up. The explorer url has the resourceRoot=./. Any ideas why explorer is not showing the new languages?

**Comment by gjcope** [comment]
In the thread above, it looks like @dirkroorda added the languages to the distributed js file for Voyager Story. The same changes would need to be added to Voyager Explorer as well. To avoid that, you could make the changes in the source and build all the versions at once.

**Comment by dirkroorda** [comment]
Yes, I made those changes in my version of voyager-story that I use in the pilot `voyager-story`, to be precise, the file voyager-story.dev.patch.js in https://github.com/CLARIAH/pure3d/tree/main/pilots/static/dist/js.

@vicding-mi and @kgschoueri are working with different instances of the voyager.

**Comment by kgschoueri** [comment]
Yes, @vicding-mi and I will try tomorrow to see what is happening with the explorer version of the new languages. We did put the new language files into the appropriate places in our source code. Maybe it has to do with the cdn.jsdelivr. On the HTML page of Voyager Explorer it is still pulling from this source whereas with Story it is pulling from cdnjs.cloudfare. 

EXPLORER

```
<!DOCTYPE html>
<html>
<head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">

    <title>Voyager Explorer v0.21.1  DEV</title>


        <link rel="shortcut icon" type="image/png" href="https://cdn.jsdelivr.net/gh/smithsonian/dpo-voyager@latest/assets/favicon.png"/>
        <link href="https://cdn.jsdelivr.net/gh/smithsonian/dpo-voyager@latest/assets/fonts/fonts.css" rel="stylesheet">

    <style>
        body { background: #303030; }
    </style>
<script defer src="[js/voyager-explorer.dev.js](https://viewer.acc.pure3d.eu/js/voyager-explorer.dev.js)"></script></head>
<body>
<voyager-explorer></voyager-explorer>
</body>
</html>
```

STORY 

```
<!DOCTYPE html>
<html>
    <head>
        <meta charset="utf-8">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">

        <title>Voyager Story v0.21.1  DEV</title>

        <link rel="shortcut icon" type="image/png" href="[favicon.png](https://viewer.acc.pure3d.eu/favicon.png)"/>
        <link href="[/fonts/fonts.css](https://viewer.acc.pure3d.eu/fonts/fonts.css)" rel="stylesheet">

        <style>
            body { background: #303030; }
        </style>

        <link href="https://cdnjs.cloudflare.com/ajax/libs/quill/1.3.7/quill.snow.css" rel="stylesheet">
        <script src="https://cdnjs.cloudflare.com/ajax/libs/quill/1.3.7/quill.js"></script>

    <script defer src="[js/voyager-story.dev.js](https://viewer.acc.pure3d.eu/js/voyager-story.dev.js)"></script><link href="[css/voyager-story.dev.css](https://viewer.acc.pure3d.eu/css/voyager-story.dev.css)" rel="stylesheet"></head>
    <body>
        <voyager-story></voyager-story>
    </body>
</html>`
```

**Comment by gjcope** [comment]
Those two links are just referencing the favicon and fonts, so the difference there doesn't affect the loading of language files. @vicding-mi Did you make the language updates to the source and then rebuild both Voyager Story and Voyager Explorer, or were the changes just made to voyager-story.dev.js? If both versions (Explorer and Story) were rebuilt, the behavior will be the same as they are using the same code path for loading.

**Comment by gjcope** [comment]
If you can give me access to an example, or a remote session should work. My guess is that you are not building the correct version but it's hard to say without seeing it. Changing the resourceRoot on the component should have an effect.

**Comment by vicding-mi** [comment]
hi Jamie @gjcope I just pump up the docker image to voyager 0.22.0. There are some new/old issues:
1. When **click on the model** (important for reproducing the error) and switch language, it throws an 'Unhandled Promise Rejection: NotFound: The object can not be found here' Please see the attached screenshot
2. the `resourceRoot="./"` is added to the html file, but not effective, symptom is voyager still trying to load files through CDN. I am sure this time I did the recompile ;) Adding resourceRoot to URL still works, this time for both story and explorer

We could wait till you merge the PR, then we do not have to worry about resourceRoot. It's more for the future if we want some new features added to voyager locally before your review/merge. 

![Screenshot 2022-10-13 at 11 27 48](https://user-images.githubusercontent.com/31689867/195565150-0a32a2ea-7074-4612-a542-b1ab82ec22bf.png)

**Comment by gjcope** [comment]
Hi @vicding-mi 

1. This is a bug that is fixed in the dev-bugfix branch and will be included for the next minor release.
2. I'm still not able to reproduce this. Is there an example you can share?

**Comment by vicding-mi** [comment]
Hi @gjcope , I just pull in the master branch which has languages merged. However, I do see the languages resources popup on the container while they are not available through the CDN yet. Is there a delay for deploy files to CDN? 

And the error I mentioned in the previous message persists. If you want, we can have a quick remote session to see it. BTW, which version of npm/node are you using? I am running `node:17-alpine`

**Comment by gjcope** [comment]
@vicding-mi The CDN pulls from published releases. We need to finish up our internal testing before release. One the release is pushed there should be no delay.

I have reproduced issue 2. above and will include the fix in this release.

**Comment by gjcope** [comment]
Closing this as I believe all the issues in this thread have been addressed in v0.22.1.
